Implement color scheme badge artifact workaround.

johnwchadwick opened the pull request:
Color scheme badges have a small gap artifact caused by an interplay between anti-aliasing and alpha blending. Here, we use pseudo-classes to ensure one opaque pixel of intersection between the shadow border and the color badge.

### Explanation
For an explanation on _why_ this is necessary, please [see this codepen](https://codepen.io/john-chadwick/pen/mdmmWKG).

### Preview
This fix is not entirely perfect; a more rigorous fix might involve putting the box shadow itself into a pseudo-element and making it overlap that way. (Since box-shadows are always clipped by the shape of the element, this would be necessary to work around the problem.)
![SHAIHRK2z2](https://user-images.githubusercontent.com/86682572/126001016-6789a53c-982c-4335-89c4-7aeb451e030d.gif)

johnwchadwick committed:
Implement color scheme badge artifact workaround.

Color scheme badges have a small gap artifact caused by an interplay
between anti-aliasing and alpha blending. Here, we use pseudo-classes to
ensure one opaque pixel of intersection between the shadow border and
the color badge.

reynolek commented on packages/insomnia-app/app/ui/components/settings/theme-panel.tsx:
```diff
@@ -45,6 +44,19 @@ const ThemeButton = styled.div<{ $isActive: boolean; $isInOsThemeMode: boolean }
   } : {}),
+  // This is a workaround for some anti-aliasing artifacts that impact the color scheme badges.
+  ...($isActive && $isInOsThemeMode ? {
+    '&:not(:hover):before': {
```
![magic](https://media.giphy.com/media/12NUbkX6p4xOO4/giphy.gif)

develohpanda reviewed (approved):
That codepen blew me away. Nice one, thanks for fixing this! 🙌🏽 @dimitropoulos I can't seem to find whether a Linear ticket was created for this issue?

johnwchadwick committed:
Improved workaround for color badge bug.

This improved workaround has fewer limitations and reduces visual
artifacts during transitions.

johnwchadwick commented:
Apologies, but I've re-requested review. There were some visual glitches during transitions that were bugging me enough to revisit this and implement the proposed better solution in the original comment (putting the actual box shadow on the pseudo element.)

![Animation of improved workaround.](https://user-images.githubusercontent.com/86682572/126205462-32574219-bb99-457c-aad9-98ef8eaaede5.gif)

This fixes several artifacts:

- The noticeable jump when switching between OS color scheme mode and back.
- The workaround border being visible during theme transitions, as it was not synchronized with the rest of the animation. There is no longer a separate 'workaround' border.
- The desync between the color badge and the outline when switching themes. This one can actually be fixed without this, but it still was bugging me.

It is still not totally perfect, but at least the remaining artifacts are _a lot_ less noticeable.

dimitropoulos reviewed (approved):
tested it again, looks and works great!

johnwchadwick committed:
Don't fade ring when hovering in OS theme mode.

This tiny change makes the transition look a lot better.

johnwchadwick commented:
OK, OK! One last minor tweak.

The animation when hovering looks a little bit janky. To illuminate, this is what it looks like in 25% speed:

![Ring hover animation before](https://user-images.githubusercontent.com/86682572/126212709-f7915b82-6ab6-40d2-8353-d277ac197085.gif)

This happens because the ring fades out, but it would probably be better if it only transitioned in spread and not color. In my opinion, it's a pretty big gain for a one-line tweak.

![Ring hover animation after](https://user-images.githubusercontent.com/86682572/126212858-1e85328c-268a-4e0d-89f7-c525637b0b1f.gif)

The one-line tweak is that the `:hover:before` rule now specifies a zero-width `box-shadow` instead of `none`.

dimitropoulos reviewed (approved):
(that last change was my fault, haha, we got talking about it while waiting for the CI before merging and it's so easy to fix that it was worth waiting for the CI one last time).

develohpanda commented:
This is amazing @johnwchadwick 😍
